dev/src/transaction.ts

```diff
@@ -622,17 +619,124 @@ export class Transaction implements firestore.Transaction {
+      // response because we are not starting a new transaction
+      return this._transactionIdPromise.then(async opts => {
+        const r = await resultFn.call(this, param, opts);
+        return r.result;
+      });
+    } else {
+      if (this._readOnlyReadTime) {
+        // We do not start a transaction for read-only transactions
+        // do not set _prevTransactionId
+        return resultFn
+          .call(this, param, this._readOnlyReadTime)
+          .then(resolveResult);
+      } else {
+        // This is the first read of the transaction so we create the appropriate
+        // options for lazily starting the transaction inside this first read op
+        let opts: api.ITransactionOptions;
+        if (this._writeBatch) {
+          opts = this._prevTransactionId
+            ? {readWrite: {retryTransaction: this._prevTransactionId}}
+            : {readWrite: {}};
+        } else {
+          opts = {readOnly: {}};
+        }
+
+        const resultPromise = resultFn.call(this, param, opts);
+
+        // Ensure the _transactionIdPromise is set synchronously so that
+        // subsequent operations will not race to start another transaction
+        this._transactionIdPromise = resultPromise.then(async r => {
+          if (!r.transaction) {
+            // Illegal state
+            // The read operation was provided with new transaction options but did not return a transaction ID
+            // Rejecting here will cause all queued reads to reject
+            throw new Error('Transaction ID was missing from server response');
+          }
+          return r.transaction;
+        });
+
+        return resultPromise.then(resolveResult);
+      }
+    }
+  }
+
+  private async getSingleFn<
+    AppModelType,
+    DbModelType extends firestore.DocumentData,
+  >(
+    document: DocumentReference<AppModelType, DbModelType>,
+    opts: Uint8Array | api.ITransactionOptions | Timestamp
+  ): Promise<{
+    transaction?: Uint8Array;
+    result: DocumentSnapshot<AppModelType, DbModelType>;
+  }> {
+    const documentReader = new DocumentReader(this._firestore, [document]);
+    if (opts instanceof Timestamp) {
+      documentReader.readTime = opts;
+    } else {
+      documentReader.transactionIdOrNewTransaction = opts;
+    }
+    const {
+      transaction,
+      result: [result],
+    } = await documentReader.getResponse(this._requestTag);
+    return {transaction, result};
+  }
+
+  private async getBatchFn<
+    AppModelType,
+    DbModelType extends firestore.DocumentData,
+  >(
+    {
+      documents,
+      fieldMask,
+    }: {
+      documents: Array<DocumentReference<AppModelType, DbModelType>>;
+      fieldMask?: FieldPath[] | null;
+    },
+    opts: Uint8Array | api.ITransactionOptions | Timestamp
+  ): Promise<{
+    transaction?: Uint8Array;
+    result: DocumentSnapshot<AppModelType, DbModelType>[];
+  }> {
+    const documentReader = new DocumentReader(this._firestore, documents);
```

Comment:
I think the DocumentReader constructor should take `opts` as a third parameter. If you want, you can do this refactor.

Reply:
Hmm then wouldn't this also be the case for `readTime` and `fieldMask`? I do kind of think that all of these should actually be immutable options provided at construction rather than mutable fields that could be changed at any point before/during/after the request... Do you want me to change all of them to be provided in a single `opts` constructor parameter?

```
opts?: {
      transactionOrReadTime?: Uint8Array | api.ITransactionOptions | Timestamp;
      fieldMask?: FieldPath[];
    }
```

Reply:
I completely agree that with making this immutable passed through constructor. Whether this should be an `opts` object or have `transactionOrReadTime` and `fieldMask` be parameters, I am inclined to go the parameter route, since that is the pattern used most other places in code. Regardless, what you implemented is a great improvement over previous code IMHO.